src/widgets/tooltip.rs

```diff
@@ -0,0 +1,213 @@
+use std::sync::{Arc};
+use crate::core::{
+    Bound, Children, Color, EventType, MutableBound, OnEvent, rsx, widget,
+    render_command::RenderCommand, 
+    styles::{PositionType, Style, StyleProp, Units}
+};
+
+use crate::widgets::{Background, Clip, Element, If, Text};
+
+#[derive(Clone, PartialEq, Debug, Default)]
+pub struct TooltipData {
+    /// The anchor coordinates in pixels (x, y)
+    ///
+    /// If `None`, defaults to cursor position
+    pub anchor: Option<(f32, f32)>,
+    /// Used to force an update
+    ticker: u8,
+    /// The size of the tooltip in pixels (width, height)
+    pub size: Option<(f32, f32)>,
+    /// The text to display
+    pub text: String,
+    /// Whether the tooltip is visible or not
+    pub visible: bool,
+}
+
+impl TooltipData {
+    /// Force the tooltip to re-render
+    pub fn mark_dirty(&mut self) {
+        // Hack for forcing the Binding to re-render
+        self.ticker = self.ticker.wrapping_add(1);
+    }
+}
+
+
+/// A provider for managing a tooltip.
+///
+/// This widget creates a single tooltip that can be controlled by any descendant [TooltipConsumer].
+///
+/// # Arguments
+///
+/// * `position`: The position of the containing rect (used to layout the tooltip).
+/// * `size`: The size of the containing rect (used to layout the tooltip).
+///
+/// # Examples
+///
+/// ```
+/// # use kayak_ui::core::{rsx, widget};
+///
+/// #[widget]
+/// fn MyWidget() {
+///   rsx! {
+///     <>
+///         <TooltipProvider size={Some(1280.0, 720.0)}>
+///             // ...
+///             <TooltipConsumer text={"Tooltip A".to_string()}>
+///                 // ...
+///             </TooltipConsumer>
+///             <TooltipConsumer text={"Tooltip B".to_string()}>
+///                 // ...
+///             </TooltipConsumer>
+///             // ...
+///         </TooltipProvider>
+///     </>
+///   }
+/// }
+/// ```
+#[widget]
+pub fn TooltipProvider(
```

Comment:
Should tooltips be removed as a widget or are they okay to keep?

There's a few things they need to be _mostly_ feature-complete:
1. Dynamic resizing based on text (is there a way to get this info on render?)
2. Nesting of consumers (hover events are called on every node that the cursor is hovered on, whether on top or not, preventing this from working)
3. Some layout issues regarding `right` and `bottom` (might be an OS issue?)

They're good enough to work with, but no where near what they should be as of yet.

Reply:
> Should tooltips be removed as a widget or are they okay to keep?

They are fine to keep!

> Dynamic resizing based on text (is there a way to get this info on render?)

Not yet but I will be adding a measure_text possibly to the context?

> Nesting of consumers (hover events are called on every node that the cursor is hovered on, whether on top or not, preventing this from working)

We can perhaps add a `hoverable` similarly to how we handled `focusable`?

> Some layout issues regarding right and bottom (might be an OS issue?)

Layout issues with text?

Reply:
> Not yet but I will be adding a measure_text possibly to the context?

That would be very helpful. Is there any plans to allow widgets to access their cached layout? Sorta like [`onLayout`](https://reactnative.dev/docs/view#onlayout) from React Native.

> We can perhaps add a `hoverable` similarly to how we handled `focusable`?

Maybe. This would help solve the issue of what's being hovered. Although, could we also achieve this by adding a way to set `should_propagate` to `false`? Then we can propagate it like `Text` → `Clip` → `Background` → `MyWidget` and handle the event on `MyWidget` (or just prevent it from bubbling up to the parent).

But I think we also need a way to only mark only the topmost node as "hovered". If a window is on top of another window, both containing buttons, then both buttons might be marked as "hovered" when really we only want the button from the topmost window. Is there a way we could do this?

I also may have no clue what I'm talking about haha

> Layout issues with text?

No, with the containing `Background` (and all other widgets). Maybe I'm doing something wrong, but `right` and `bottom` position the tooltip stuck to the edge of the screen. And (maybe related) `padding_*`, and `margin_*` seem to not work at all for me.

I thought it might be an OS issue or another DPI issue because my widgets are all cutoff at 50% width and height of the physical window. So maybe that's also affecting layout? I'll open an issue about it later tonight with more details.

Reply:
> No, with the containing Background (and all other widgets). Maybe I'm doing something wrong, but right and bottom position the tooltip stuck to the edge of the screen. And (maybe related) padding_*, and margin_* seem to not work at all for me.

Ah yeah I think I have to remove margin I was miss-understanding as to how morphorm works with margins. You have to use left, right, top, bottom to add margin.

Reply:
Oh okay gotcha. That's what I ended up doing anyways so that works out! 😁